Open Judge System/Web/OJS.Web/Areas/Contests/Controllers/ResultsController.cs

```diff
@@ -360,33 +360,69 @@ public ActionResult Stats(int contestId, bool official)
 
             var fromPoints = 0;
             var toPoints = 0;
-            foreach (var problem in contestResults.Problems)
+
+            if (contest.IsOnline && official)
+            {
+                statsModel.IsGroupedByProblemGroup = true;
+
+                foreach (var problems in contestResults.Problems.GroupBy(p => p.ProblemGroupId))
+                {
+                    var maxResultsForProblemGroupCount = contestResults.Results
+                        .Count(r => r.ProblemResults?
+                            .Any(pr => problems
+                                .Any(p =>
+                                    p.Id == pr.ProblemId &&
+                                    pr.BestSubmission.Points == pr.MaximumPoints)) ?? false);
+
+                    var maxPointsForProblemGroup = problems.FirstOrDefault()?.MaximumPoints ?? default(int);
+                    var problemGroupName = string.Join("<br/>", problems.Select(p => p.Name));
+
+                    AddStatsByProblem(problemGroupName, maxPointsForProblemGroup, maxResultsForProblemGroupCount);
+                    AddStatsByPointsRange(maxPointsForProblemGroup, contestResults.Results);
+                }
+            }
+            else
+            {
+                foreach (var problem in contestResults.Problems)
+                {
+                    var maxResultForProblemCount = contestResults.Results
```

Comment:
Consider extracting in data service.

Reply:
I operate on already queried set of results from a view model, it is not for data service